consensus/hotstuff/signature/weighted_signature_aggregator_test.go

```diff
@@ -0,0 +1,261 @@
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/flow-go/consensus/hotstuff"
+	"github.com/onflow/flow-go/crypto"
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/signature"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+func sortIdentities(ids []flow.Identity) []flow.Identity {
+	canonicalOrder := func(i, j int) bool {
+		return bytes.Compare(ids[i].NodeID[:], ids[j].NodeID[:]) < 0
+	}
+	sort.Slice(ids, canonicalOrder)
+	return ids
+}
+
+func sortIdentifiers(ids []flow.Identifier) []flow.Identifier {
+	canonicalOrder := func(i, j int) bool {
+		return bytes.Compare(ids[i][:], ids[j][:]) < 0
+	}
+	sort.Slice(ids, canonicalOrder)
+	return ids
+}
+
+func createAggregationData(t *testing.T, signersNumber int) (
+	hotstuff.WeightedSignatureAggregator, []flow.Identity, []crypto.Signature, []byte, string) {
+	// create identities
+	ids := make([]flow.Identity, 0, signersNumber)
+	for i := 0; i < signersNumber; i++ {
+		ids = append(ids, *unittest.IdentityFixture())
+	}
+	ids = sortIdentities(ids)
+
+	// create message and tag
+	msgLen := 100
+	msg := make([]byte, msgLen)
+	tag := "random_tag"
+	hasher := crypto.NewBLSKMAC(tag)
+
+	// create keys, identities and signatures
+	sigs := make([]crypto.Signature, 0, signersNumber)
+	seed := make([]byte, crypto.KeyGenSeedMinLenBLSBLS12381)
+	for i := 0; i < signersNumber; i++ {
+		// keys
+		_, err := rand.Read(seed)
+		require.NoError(t, err)
+		sk, err := crypto.GeneratePrivateKey(crypto.BLSBLS12381, seed)
+		require.NoError(t, err)
+		ids[i].StakingPubKey = sk.PublicKey()
+		// signatures
+		sig, err := sk.Sign(msg, hasher)
+		require.NoError(t, err)
+		sigs = append(sigs, sig)
+	}
+	aggregator, err := NewWeightedSignatureAggregator(ids, msg, tag)
+	require.NoError(t, err)
+	return aggregator, ids, sigs, msg, tag
+}
+
+func verifyAggregate(signers []flow.Identifier, ids []flow.Identity, sig []byte,
+	msg []byte, tag string) (bool, error) {
+	// query identity using identifier
+	// done linearly just for testing
+	getIdentity := func(signer flow.Identifier) *flow.Identity {
+		for _, id := range ids {
+			if id.NodeID == signer {
+				return &id
+			}
+		}
+		return nil
+	}
+
+	// get keys
+	keys := make([]crypto.PublicKey, 0, len(ids))
+	for _, signer := range signers {
+		id := getIdentity(signer)
+		if id == nil {
+			return false, errors.New("unexpected test error")
+		}
+		keys = append(keys, id.StakingPubKey)
+	}
+	// verify signature
+	hasher := crypto.NewBLSKMAC(tag)
+	return crypto.VerifyBLSSignatureOneMessage(keys, sig, msg, hasher)
+}
+
+func TestWeightedSignatureAggregator(t *testing.T) {
+	signersNum := 20
+
+	// constrcutor edge cases
+	t.Run("constructor", func(t *testing.T) {
+		msg := []byte("random_msg")
+		tag := "random_tag"
+
+		signer := unittest.IdentityFixture()
+		// identity with empty key
+		_, err := NewWeightedSignatureAggregator([]flow.Identity{*signer}, msg, tag)
+		assert.Error(t, err)
```

Comment:
I believe here we will get a sentinel error, please add assertion that we really expect it. 
`assert.True(t, engine.IsInvalidInputError(err))`